In [82]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [83]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [84]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [85]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [86]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaled_data = StandardScaler().fit_transform(df_market_data)

In [87]:
# Create a DataFrame with the scaled data
df_scaled_market_data = pd.DataFrame(
    scaled_data,
    columns=df_market_data.columns
)

# Copy the crypto names from the original data
df_scaled_market_data["coin_id"] = df_market_data.index

# Set the coinid column as index
df_scaled_market_data = df_scaled_market_data.set_index("coin_id")

# Display sample data
df_scaled_market_data.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [88]:
# Create a list with the number of k-values from 1 to 11
k = range(1,12)
# list(k_list)

In [89]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    model = KMeans(n_clusters=i)
# 2. Fit the model to the data using `df_market_data_scaled`
    model.fit(df_scaled_market_data)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(model.inertia_)
    
inertia

C:\Users\simon\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


[287.0,
 198.57181809798126,
 123.19048183836959,
 79.02243535120975,
 63.85866780584265,
 53.28303462223397,
 45.642278653808205,
 37.284806594046614,
 32.93002674292087,
 28.165433305979256,
 24.116198421409198]

In [90]:
# Create a dictionary with the data to plot the Elbow curve
inertia_dict = {'k':k, 'inertia':inertia}

# Create a DataFrame with the data to plot the Elbow curve
inertia_df = pd.DataFrame(inertia_dict)
inertia_df

k     inertia
0    1  287.000000
1    2  198.571818
2    3  123.190482
3    4   79.022435
4    5   63.858668
5    6   53.283035
6    7   45.642279
7    8   37.284807
8    9   32.930027
9   10   28.165433
10  11   24.116198

In [91]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
inertia_df.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value of k is k=4


---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [92]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4)

In [93]:
# Fit the K-Means model using the scaled data
model.fit(df_scaled_market_data)

KMeans(n_clusters=4)

In [94]:
# Predict the clusters to group the cryptocurrencies using the scaled data
clusters = model.predict(df_scaled_market_data)

# Print the resulting array of cluster values.
clusters

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0])

In [95]:
# Create a copy of the DataFrame
df_scaled_market_data2 = df_scaled_market_data.copy()


In [96]:
# Add a new column to the DataFrame with the predicted clusters
df_scaled_market_data2['cluster_original'] = clusters

# Display sample data
df_scaled_market_data2.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  cluster_original  
coin_id                                                     
bitcoin                        -0.251637                 1  
ethereum                       -0.199352                 1  
tether                         -0.282061                 0  
ripple                         -0.295546                 0  
bitcoin-cash                   -0.270317                 1

In [97]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

plot_cluster_data = df_scaled_market_data2.hvplot.scatter(
    x     = "price_change_percentage_24h", 
    y     = "price_change_percentage_7d", 
    by    = 'cluster_original',
    title = "Original Data. K-Mean Clusters with k=4.",
    hover_cols = 'coin_id'
)

plot_cluster_data


:NdOverlay   [cluster_original]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [98]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components = 3)

In [99]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
pca_market_data = pca.fit_transform(df_scaled_market_data)

# View the first five rows of the DataFrame. 
pca_market_data[0:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [100]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
# Retrieve the explained variance to determine how much information can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

In [101]:
#****** This cell added to calculate the total explained variance of the three components *******
pca.explained_variance_ratio_.sum()

0.8950316570309841

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance is 0.895 which gives a value of 90% 

In [102]:
# Create a new DataFrame with the PCA data.
df_pca_market_data = pd.DataFrame(pca_market_data, columns=['PC1','PC2','PC3'])


# Copy the crypto names from the original data
df_pca_market_data['coin_id'] = df_scaled_market_data.index

# Set the coinid column as index
df_pca_market_data = df_pca_market_data.set_index('coin_id')

# Display sample data
df_pca_market_data.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [103]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,12))

In [104]:
# Create an empy list to store the inertia values
pca_inertia = []

# Create a for loop to compute the inertia with each possible value of k

for i in k:
    # Inside the loop:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    model_pca = KMeans(n_clusters = i)
    # 2. Fit the model to the data using `df_market_data_pca`
    model_pca.fit(df_pca_market_data)
    # 3. Append the model.inertia_ to the inertia list
    pca_inertia.append(model_pca.inertia_)

pca_inertia

C:\Users\simon\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


[256.8740855678926,
 168.8118963068523,
 93.77462568057304,
 49.6654966517974,
 37.87874703346251,
 27.720626194899825,
 21.134056037473616,
 17.34467876688018,
 13.668117456693249,
 10.484890485976928,
 8.128620320213502]

In [105]:
# Create a dictionary with the data to plot the Elbow curve
pca_inertia = {'k':k, 'pca_inertia':pca_inertia}
# Create a DataFrame with the data to plot the Elbow curve
pca_inertia_df = pd.DataFrame(pca_inertia)

pca_inertia_df = pca_inertia_df.set_index('k')
pca_inertia_df


pca_inertia
k              
1    256.874086
2    168.811896
3     93.774626
4     49.665497
5     37.878747
6     27.720626
7     21.134056
8     17.344679
9     13.668117
10    10.484890
11     8.128620

In [106]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

plot_clusterpca_data = pca_inertia_df['pca_inertia'].hvplot(
    title  = 'Principal Components Analysis PCA : Inertia versus Number of Clusters (k)',
    xlabel = 'Number of Clusters (k)',
    ylabel = 'Inertia', 
    color = 'Red'
)

plot_clusterpca_data

:Curve   [k]   (pca_inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for 'k' when using the PCA processed data is k = 4 with an inertia of 49.67.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** The best value in the original data was also considered to be k = 4 but it should be noted the inertia at this value was 79.02.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [107]:
# Initialize the K-Means model using the best value for k
k = 4
model_pca = KMeans(n_clusters=k, random_state=1)

In [108]:
# Fit the K-Means model using the PCA data
model_pca.fit(df_pca_market_data)

KMeans(n_clusters=4, random_state=1)

In [109]:
# Predict the clusters to group the cryptocurrencies using the PCA data
cluster = model_pca.predict(df_pca_market_data)

# Print the resulting array of cluster values.
cluster[0:5]

array([1, 1, 0, 0, 1])

In [110]:
# Create a copy of the DataFrame with the PCA data
df_pca_market_data2 = df_pca_market_data.copy()

# Add a new column to the DataFrame with the predicted clusters
df_pca_market_data2['pca_pr-clusters'] = cluster

# Display sample data
df_pca_market_data2.head()


PC1       PC2       PC3  pca_pr-clusters
coin_id                                                    
bitcoin      -0.600667  0.842760  0.461595                1
ethereum     -0.458261  0.458466  0.952877                1
tether       -0.433070 -0.168126 -0.641752                0
ripple       -0.471835 -0.222660 -0.479053                0
bitcoin-cash -1.157800  2.041209  1.859715                1

In [111]:
df_pca_market_data2.shape


(41, 4)

In [112]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.


# Add a new column to the DataFrame with the pca predicted clusters
df_scaled_market_data3 = df_scaled_market_data2.copy()

df_scaled_market_data3['pca_pr-clusters'] = cluster
#df_scaled_market_data3.head()

plot_cluster2_data = df_scaled_market_data3.hvplot.scatter(
    x     = "price_change_percentage_24h", 
    y     = "price_change_percentage_7d", 
    by    = 'pca_pr-clusters',
    title = "PCA Processed Data. K-Mean Clusters with k=4.",
    hover_cols = 'coin_id'
)

plot_cluster2_data



:NdOverlay   [pca_pr-clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [113]:
# Composite plot to contrast the Elbow curves

inertia_df.hvplot.line(x="k", y="inertia", title="Elbow Curve", xticks=k, width=400, xlim=(1, 11), ylim=(0, 310)) + pca_inertia_df.hvplot(title  = 'Elbow Curve : PCA', xlabel = 'k',ylabel = 'Inertia', width=400, color='Red', xlim=(1, 11), ylim=(0, 310))




:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (pca_inertia)

In [114]:
# Composite plot to contrast the clusters

plot_cluster_data + plot_cluster2_data


:Layout
   .NdOverlay.I  :NdOverlay   [cluster_original]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [pca_pr-clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

#### Answer the following question: 

  * **Question:** After visually analysing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The cluster analysis results show that the same datapoints are assigned to the same group of clusters and therefore the two processes have achieved the same result. However the PCA processed data has achieved the result in a more efficient manner.